RangeFinder: add support to TFMini Plus

lucasdemarchi opened the pull request:
[New lidar from Benewake](http://en.benewake.com/product/detail/5c345cd0e5b3a844c472329b.html) that supports I2C in addition to UART in the previous version. Besides the bus difference, commands are different enough to deserve a new driver implementation. Commands also don't follow the usual reg map in I2C, but use a similar protocol as when it's connected on the UART.

This implements it on I2C mode. A python script is provided to switch the sensor to I2C before connecting to the flight controller. This can also be done by tools provided by Benewake. The I2C address is also configurable, but still not tested. This would allow one to connect more than one sensor on the same I2C bus.

It's working fine and reliably for me, but it's not flight tested.

In order to test you have to set at least the parameters to select the I2C address and the rangefinder type:

RNGFND1_ADDR = 16
RNGFND1_TYPE = 24

bys1123 commented:
FYI, It also support lower minimum distance to 10cm.

lucasdemarchi commented:
Yes, that's already reflected in the driver, although the user also would also have to set the other params:

RNGFND1_MIN_CM = 10
RNGFND1_MAX_CM = 1200

lucasdemarchi commented on libraries/AP_RangeFinder/AP_RangeFinder_Benewake_TFMiniPlus.cpp:
```diff
@@ -0,0 +1,214 @@
+    hal.scheduler->delay(100);
+
+    ret = _dev->transfer(nullptr, 0, val, 7);
+    if (!ret || val[0] != 0x5A || val[1] != 0x07 || val[2] != 0x01) {
```
firmware is not fixed. It's **at least** 1.7.6. Mine, for example is 1.9.something.

but I guess I could add a checksum check nonetheless.

lucasdemarchi replied:
checksum added

lucasdemarchi commented on libraries/AP_RangeFinder/AP_RangeFinder_Benewake_TFMiniPlus.cpp:
```diff
@@ -0,0 +1,214 @@
+            hal.console->printf(DRIVER ": Unable to set configuration register %u\n",
+                                cmds[i][2]);
+            hal.scheduler->delay(100);
+            goto fail;
```
Those values are actually saved in the sensor. I didn't check the response as it didn't seem really important.... I'll double check later if it would make sense.

WickedShell commented on libraries/AP_RangeFinder/AP_RangeFinder_Benewake_TFMiniPlus.cpp:
```diff
@@ -0,0 +1,214 @@
+    uint8_t val[12], i;
+
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        hal.console->printf(DRIVER ": Unable to get bus semaphore\n");
```
Isn't this always dead code? Why not just use `_dev->get_semaphore()->take_blocking()` which doesn't require the surrounding `if` section?

lucasdemarchi replied:
fixed

WickedShell commented on libraries/AP_RangeFinder/AP_RangeFinder_Benewake_TFMiniPlus.cpp:
```diff
@@ -0,0 +1,214 @@
+        if (!ret) {
+            hal.console->printf(DRIVER ": Unable to set configuration register %u\n",
+                                cmds[i][2]);
+            hal.scheduler->delay(100);
```
`delay` makes sense in debug prints when trying to ensure the console gets to print before going into bad code, but why include it in the normal firmware?

lucasdemarchi replied:
The delay is actually in the wrong place. It should be there in the success case, not on failure.

lucasdemarchi replied:
I'll check if we really need the delay

lucasdemarchi replied:
fixed

lucasdemarchi commented:
Rebased and fixed comments. New parameters, since uavcan's rangefinder was applied:

```
RNGFND1_ADDR = 16
RNGFND1_TYPE = 25
RNGFND1_MIN_CM = 10
RNGFND1_MAX_CM = 1200
```

bys1123 commented:
I'm testing this driver today, but I can't get read from this sensor.

So I connected the SCL SDA to my logic analyzer tool.
I'm setting RNGFND1_ADDR = 16
![image](https://user-images.githubusercontent.com/4748411/58750803-3391d900-84c9-11e9-85e4-7a7c41146bd5.png)
But it's actually requesting 0x20.
![image](https://user-images.githubusercontent.com/4748411/58750863-c894d200-84c9-11e9-9bec-56eb16f4f8eb.png)
And the TFminiPlus default I2C address is 0x10
![image](https://user-images.githubusercontent.com/4748411/58750872-fe39bb00-84c9-11e9-99dd-1cce0601bb19.png)

@lucasdemarchi

lucasdemarchi commented:
16 in decimal is 0x10 in hexa. You are seeing 0x20 because your logical analyzer is not shifting the value for the read/write bit. There's probably a configuration to change that in your analyzer.

It's not working for you because of the missing stop condition. You need #11405 to fix that in your board.

bys1123 commented:
@lucasdemarchi Thanks for the information. I'll test #11405 later.
My analyzer can recognize the R/W bit I remember, but I don't know what happened on this, I'll check later.

lucasdemarchi commented:
It can recognize (see it shows as data write), but you have to tell it to shift the address value to ignore it on the address. Or just know that the value is correct because of that :)

bys1123 commented on Tools/scripts/tfminiplus.py:
```diff
@@ -0,0 +1,78 @@
+
+SYSTEM_RESET  = struct.pack('B' * 4, 0x5A, 0x04, 0x02, 0x60)
+MODE_I2C      = struct.pack('B' * 5, 0x5A, 0x05, 0x0A, 0x01, 0x6A)
+MODE_UART     = struct.pack('B' * 5, 0x5A, 0x05, 0x0A, 0x01, 0x69)
```
Here is a typo, I fixed in #11459

bys1123 commented:
![image](https://user-images.githubusercontent.com/4748411/58761934-3bf71c00-857c-11e9-8f57-b3c909a76998.png)
@lucasdemarchi I tested #11405, still not working, other driver has NACK now, but not this one.
![image](https://user-images.githubusercontent.com/4748411/58761965-b58f0a00-857c-11e9-86ea-d31045989e08.png)
And you can see the R/W bit is recognized now. Don't know why this tfmini driver is using Data write, not Address write.

lucasdemarchi commented:
You still don't have the stop in the first configuration commands. What board is that? Did you update the submodules and do a clean build when testing that PR?

bys1123 commented:
Sorry, my fault, checked wrong branch after fetch #11405.

bys1123 commented:
![image](https://user-images.githubusercontent.com/4748411/58763421-0e669e80-858d-11e9-823f-365355c1a061.png)
Tested on Pixhawk4mini, works awesome!
